src/MainDriver.cpp

```diff
@@ -225,7 +225,30 @@ class InputProvider {
 public:
     virtual ~InputProvider() {}
     virtual FILE* getInputStream() = 0;
+    virtual bool readInput(std::string& Buf) {
+        if (Stream == nullptr) {
```

Comment:
`readInput()` should call `getInputStream()` to get a stream, and call `endInput()` before returning.